dev-docs/RFCs/v2.1/normalize-3d-tile-loaders.md

```diff
@@ -0,0 +1,396 @@
+- **Tileset Traversal** - Dynamically loading and rendering 3D tiles based on current viewing position, possibly triggering loads of new tiles and unloading of older, no-longer visible tiles.
+
+## Modules Structure
+
+**Current structure**
+
+`@loaders.gl/3d-tiles`: Load Cesium 3D tiles
+
+- loaders
+  - `Tileset3DLoader`
+  - `Tile3DHeaderLoader`
+- Tile Classes
+  - `Tileset3D:`
+  - `Tileset:`
+- Shared Components
+
+  - `BaseTileset3DTraversal`
+  - `Tileset3DCache`
+  - `RequestScheduler`
+
+`@loaders.gl/i3s`: Load ArcGIS I3S tiles
+
+- loaders
+  - `Tileset3DLoader`
+  - `Tile3DHeaderLoader`
+- Tile Classes
+  - `I3STileset`
+  - `I3STileHeader`
+- Dependencies:
+  - `@loaders.gl/3d-tiles`: `BaseTileset3DTraversal`, `Tileset3DCache`, `RequestScheduler`, etc.
+
+`@loaders.gl/potree`: Load Potree tiles
+
+- loaders
+  - `PotreeLoader`
+  - `PotreeHierarchyChunkLoader`
+  - `PotreeBinLoader`
+- Tile Classes
+  - `Octree`
+
+**Proposed structure**
+
+Each specification will have its own loader module and expose a Tileset loader, a TileHeader loader and a tile content loader. And a separate module `@loaders.gl/tiles` will contain all the tile classes and common components shared by different tile loaders.
+
+```
+|--`@loaders.gl/3d-tiles`         // Load Cesium 3D tiles
+|   |- `Tiles3DTilesetLoader`
+|   |- `Tiles3DTileLoader`
+|   |- `Tiles3DContentLoader`
+|--`@loaders.gl/i3s`              // Load ArcGIS I3S tiles.
+|   |-- `I3STilesetLoader`
+|   |-- `I3STileLoader`
+|   |-- `I3SContentLoader`
+|--`@loaders.gl/potree`           // Load Potree tiles
+|   |-- `PotreeTilesetLoader`
+|   |-- `PotreeTileLoader`
+|   |-- `PotreeContentLoader`
+|--`@loaders.gl/tiles`            // A modules contains the tile classes and common components for loading 2d and 3d tiles.
+    |-- `Tileset2D`
+    |-- `Tileset3D`
+    |-- `Tile2D`
+    |-- `Tile3D`
+    |-- `Tileset3DTraversal`
+    |-- `Tileset3DCache`
+    |-- `RequestScheduler`
+```
+
+**For users**
+
+- `Tileset2D`: A class which help manage 2d tiles.
+- `Tileset3D`: A class which can understand unified tileset format loaded from tileset loaders, i.e. `Tiles3DTilesetLoader` for `@loaders.gl/3d-tiles`, `I3STilesetLoader` for `@loaders.gl/i3s`. Also it provides helper functions to dynamically loading and unloading tiles for rendering under current viewport.
+
+**For shared by tiles loaders**
+
+- `Tile2D`: A class that extended from 2D tile data.
+- `Tile3D`: A class that can understand unified tile format loaded from tile loaders, i.e. `Tiles3DTileLoader` for `@loaders.gl/3d-tiles`, `I3STileLoader` for `@loaders.gl/i3s`.
+- `TilesetTraversal`: A class that traverse tile tree to get renderable tiles based on certain state (i.e. viewport), may cause loading and unloading tiles, and updating tiles visibilties.
+- `TilesetCache`: LRU cache for caching loaded tiles.
+- `RequestScheduler`: Managing fetching requests based on traversal priorities.
+
+```
```

Comment:
How does `Tileset3D` know which loaders to use to load the tile contents? Per module dependency hierarchy described above, the `@loaders.gl/tiles` module should not depend on any spec implementations. Is the user supposed to register the `*TileLoader` and `*TileContentLoader` as well? This is not apparent from the code sample in this RFC.

If the usage of `*TileLoader` and `*TileContentLoader` are hidden from the user, does it still make sense to have so many loaders? It is possible for a single loader to handle multiple file formats. The parser can make the switch based on url, as well as content sniffing. It will be a lot less confusing if each module only exports one loader, e.g. `Tiles3DLoader` or `I3SLoader`.

Reply:
I agree that the multiple loaders could be confusing to the user. But normally the tile header contains metadata and a `url` pointing to the actual payload, which is the `tile.content` in this proposal. 

Another way, each loader could expose 
- a `TileLoader` which accept options: `loadContent`
- a parser to parse the content